packages/transaction-controller/src/utils/history.test.ts

```diff
@@ -0,0 +1,372 @@
+
+describe('History', () => {
+  describe('updateTransactionHistory', () => {
+    it('does nothing if the history property is missing', () => {
+      const mockTransaction = createMockTransaction();
+      const originalInputTransaction = cloneDeep(mockTransaction);
+
+      const updatedTransaction = updateTransactionHistory(
+        mockTransaction,
+        'test update',
+      );
+
+      expect(updatedTransaction).toBe(mockTransaction);
+      expect(updatedTransaction).toStrictEqual(originalInputTransaction);
+    });
+
+    it('does nothing if there have been no changes', () => {
+      const originalTransaction = createMockTransaction();
+      const mockTransaction = createMockTransaction({ originalTransaction });
+      const mockTransactionClone = cloneDeep(mockTransaction);
+
+      const updatedTransaction = updateTransactionHistory(
+        mockTransaction,
+        'test update',
+      );
+
+      expect(updatedTransaction).toBe(mockTransaction);
+      expect(updatedTransaction).toStrictEqual(mockTransactionClone);
+    });
+
+    it('adds a new history entry', () => {
+      const originalTransaction = createMockTransaction();
+      const mockTransaction = createMockTransaction({
+        originalTransaction,
+        partialTransaction: {
+          history: [originalTransaction],
+          txParams: { from: generateAddress(123) },
+        },
+      });
+
+      const updatedTransaction = updateTransactionHistory(
+        mockTransaction,
+        'test update',
+      );
+
+      expect(updatedTransaction).not.toBe(mockTransaction);
+      expect(updatedTransaction).toStrictEqual({
+        ...mockTransaction,
+        history: [
+          originalTransaction,
+          [
+            {
+              note: 'test update',
+              op: 'replace',
+              path: '/txParams/from',
+              timestamp: expect.any(Number),
+              value: generateAddress(123),
+            },
+          ],
+        ],
+      });
+    });
+
+    describe('when history is past max size with non-displayed entries', () => {
+      it('merges a non-displayed entry when adding a new entry after max history size is reached', () => {
+        const originalTransaction = createMockTransaction();
+        const mockTransaction = createMockTransaction({
+          partialTransaction: {
+            history: generateMockHistory({
+              originalTransaction,
+              length: MAX_HISTORY_LENGTH,
+            }),
+            // This is the changed value
+            txParams: { from: generateAddress(123) },
+          },
+        });
+        // Validate that last history entry is correct
+        // Assertion used because we know the history is set here.
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        const mockHistoryLength = mockTransaction.history!.length;
+        expect(mockTransaction.history?.[mockHistoryLength - 1]).toStrictEqual([
```

Comment:
Minor, but do we need this if it's ultimately testing a test helper?

Could we at least create a dedicated test for it so it's not duplicated in each test?

Reply:
Ah, so this is something I added for myself because I was having trouble telling what the test I wrote did 😅 . It wasn't meant to test the helper function exactly, just to check that I had called it correctly. I had more than one off-by-one error in writing this. I added this validation so that I could read the piece of the history relevant to the test, the very last one.

Reply:
This is noisy though. I'll remove these non-null assertions at least, those are bothersome.

Reply:
Alright it's a bit less noisy in 3a9cee7e, let me know what you think

Reply:
Ok, fixed in 441acde